packages/messaging/src/controllers/window-controller.ts

```diff
@@ -53,30 +67,137 @@ export class WindowController implements FirebaseMessaging, FirebaseService {
     return this.firebaseDependencies.app;
   }
 
-  async getToken(): Promise<string> {
-    if (!this.vapidKey) {
-      this.vapidKey = DEFAULT_VAPID_KEY;
+  private async messageEventListener(event: MessageEvent): Promise<void> {
+    const internalPayload = event.data as MessagePayloadInternal;
+
+    if (!internalPayload.isFirebaseMessaging) {
+      return;
+    }
+
+    // onMessageCallback is either a function or observer/subscriber.
+    if (
+      this.onMessageCallback &&
+      internalPayload.messageType === MessageType.PUSH_RECEIVED
+    ) {
+      if (typeof this.onMessageCallback === 'function') {
+        this.onMessageCallback(externalizePayload(internalPayload));
+      } else {
+        this.onMessageCallback.next(externalizePayload(internalPayload));
+      }
+    }
+
+    const dataPayload = internalPayload.data;
+    if (
+      isConsoleMessage(dataPayload) &&
+      dataPayload[CONSOLE_CAMPAIGN_ANALYTICS_ENABLED] === '1'
+    ) {
+      await this.logEvent(internalPayload.messageType!, dataPayload);
     }
+  }
+
+  getVapidKey(): string | null {
+    return this.vapidKey;
+  }
 
-    const swRegistration = await this.getServiceWorkerRegistration();
+  getSwReg(): ServiceWorkerRegistration | undefined {
+    return this.swRegistration;
+  }
 
-    // Check notification permission.
+  async getToken(options?: {
+    vapidKey?: string;
+    serviceWorkerRegistration?: ServiceWorkerRegistration;
+  }): Promise<string> {
     if (Notification.permission === 'default') {
-      // The user hasn't allowed or denied notifications yet. Ask them.
       await Notification.requestPermission();
     }
 
     if (Notification.permission !== 'granted') {
       throw ERROR_FACTORY.create(ErrorCode.PERMISSION_BLOCKED);
     }
 
-    return getToken(this.firebaseDependencies, swRegistration, this.vapidKey);
+    await this.updateVapidKey(options?.vapidKey);
+    await this.updateSwReg(options?.serviceWorkerRegistration);
+
+    if (!this.swRegistration) {
+      console.debug(
+        TAG +
+          'no sw has been provided explicitly. Attempting to find firebase-messaging-sw.js in default directory.'
+      );
+      await this.registerDefaultSw();
```

Comment:
Should we move it into `this.updateSwReg()` to be consistent with `updateVapidKey()`?
We assign the default vapid key to `this.vapidKey` inside `updateVapidKey()`. Similarly can we do the same for `this.swRegistration`?